contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,226 @@
+
+```js
+{ type: 'ADD_TODO', text: 'Go to swimming pool' }
+{ type: 'TOGGLE_TODO', index: 1 }
+{ type: 'SET_VISIBILITY_FILTER', filter: 'SHOW_ALL' }
+```
+
+### Reducer
+
+The current state, and an action is passed into a **reducer**, a pure function which returns the next state of the application
+
+In a real-world application, instead of a single reducer which can get complex, we can have multiple, simpler reducers, and combine them into one reducer.
+
+For example, we have two simpler reducers.
+
+Firstly, a reducer for the visibility filter `visibilityFilter`:
+
+```js
+function visibilityFilter(state = 'SHOW_ALL', action) {
+  if (action.type === 'SET_VISIBILITY_FILTER') {
+    return action.filter
+  } else {
+    return state
+  }
+}
+```
+
+And secondly, a reducer, that manages the todo-list items `todos`:
+
+```js
+function todos(state = [], action) {
+  switch (action.type) {
+    case 'ADD_TODO':
+      return state.concat([{ text: action.text, completed: false }])
+    case 'TOGGLE_TODO':
+      return state.map((todo, index) =>
+        action.index === index
+          ? { text: todo.text, completed: !todo.completed }
+          : todo
+      )
+    default:
+      return state
+  }
+}
+```
+
+These two reducers are called by a main reducer, and with that the complete app state is managed.
+
+```js
+function todoApp(state = {}, action) {
+  return {
+    todos: todos(state.todos, action),
+    visibilityFilter: visibilityFilter(state.visibilityFilter, action)
+  }
+}
+```
+
+The reducers are **pure** functions. They take the previous state and an action, and return the next state, as new state objects, instead of mutating the previous state.
+
+A careful eye would notice that the above code block examples on reducers are not pure function. In fact, with some Redux API, a more proper example would then be the code block below. Don't worry if you don't fully understand the specific Javascript syntax and methods used (such as `.map` and the spread operator), but just understand that new state objects are returns when state is updated.
+
+```js
+function visibilityFilter(state = 'SHOW_ALL', action) {
+  switch (action.type) {
+    case 'SET_VISIBILITY_FILTER':
+      return action.filter
+    default:
+      return state
+  }
+}
+function todos(state = [], action) {
+  switch (action.type) {
+    case 'ADD_TODO':
+      return [
+        ...state,
+        {
+          text: action.text,
+          completed: false
+        }
+      ]
+    case 'COMPLETE_TODO':
+      return state.map((todo, index) => {
+        if (index === action.index) {
+          return Object.assign({}, todo, {
+            completed: true
+          })
+        }
+        return todo
+      })
+    default:
+      return state
+  }
+}
+import { combineReducers, createStore } from 'redux'
+const reducer = combineReducers({ visibilityFilter, todos })
+const store = createStore(reducer)
+```
```

Comment:
Shorten and simply this sections as much as possible and get rid of sub-headings. Eliminate as much code as possible (because the reader is not expected to take away any code from this chapter).